runtime/src/accounts_db.rs

```diff
@@ -1925,16 +1958,63 @@ impl AccountsDb {
-        purges_zero_lamports.retain(|_pubkey, (account_infos, _ref_count)| {
-            for (_slot, account_info) in account_infos.iter() {
+        purges_zero_lamports.retain(|pubkey, (slot_account_infos, _ref_count)| {
+            // Only keep purges_zero_lamports where the entire history of the account in the root set
+            // can be purged. All AppendVecs for those updates are dead.
+            for (_slot, account_info) in slot_account_infos.iter() {
                 if store_counts.get(&account_info.store_id).unwrap().0 != 0 {
                     return false;
                 }
             }
-            true
+
+            // Exit early if not filtering more for incremental snapshots
+            if !should_filter_for_incremental_snapshots {
+                return true;
+            }
+
+            // When using incremental snapshots, do not purge zero-lamport accounts if the slot is
+            // higher than the last full snapshot slot.  This is to protect against the following
+            // scenario:
+            //
+            //      A full snapshot is taken, and it contains an account with a non-zero balance.
+            //      Later, that account's  goes to zero.  Evntually cleaning runs, and before, this
+            //      account would be cleaned up.  Finally, an incremental snapshot is taken.
+            //
+            //      Later, the incremental (and full) snapshot is used to rebuild the bank and
+            //      accounts database (e.x. if the node restarts).  The full snapshot _does_
+            //      contain the account (from above) and its balance is non-zero, however, since
+            //      the account was cleaned up in a later slot, the incremental snapshot does not
+            //      contain any info about this account, thus, the accounts database will contain
+            //      the old info from this account, which has its old non-zero balance.  Very bad!
+            //
+            // This filtering step can be skipped if there is no `last_full_snapshot_slot`, or if
+            // the `max_clean_root` is less-than-or-equal-to the `last_full_snapshot_slot`.
+            let slot_account_info_at_highest_slot = slot_account_infos
+                .iter()
+                .max_by_key(|(slot, _account_info)| slot);
+
+            slot_account_info_at_highest_slot.map_or(true, |(slot, account_info)| {
+                // Do *not* purge zero-lamport accounts if the slot is greater than the last full
+                // snapshot slot.  Since we're `retain`ing the accounts-to-purge, I felt creating
+                // the `cannot_purge` variable made this easier to understand.  Accounts that do
+                // not get purged here are added to a list so they be considered for purging later
+                // (i.e. after the next full snapshot).
+                let cannot_purge =
+                    account_info.is_zero_lamport() && *slot > last_full_snapshot_slot.unwrap();
```

Comment:
From where the `purges_zero_lamports` was constructed: https://github.com/solana-labs/solana/blob/v1.7/runtime/src/accounts_db.rs#L1811-L1819, I think it must be true that `account_info.is_zero_lamport()`, so I think we should assert that here.

Reply:
Good call. Done!